### PR TITLE
Switch logging to go-kit

### LINKED ISSUE #141: Switch logging to go-kit
About two years ago we decided to move from logrus to go-kit for logging, this is one of the repos that still needs switching. prometheus/snmp_exporter#447 is an example PR.

### PR DESCRIPTION
**Pull request by sowmiyamuthuraman**
Fixes #141

### CODE REVIEW

**Comment by brian-brazil** [comment]
Code looks good, can you add the DCO? Test failure looks like a CI issue, retrying.

**Comment by brian-brazil** [comment]
Ah, you need to update the vendoring.

**Comment by simonpasquier** [comment]
@sowmiyamuthuraman you need to update [scripts/errcheck_excludes.txt](https://github.com/prometheus/consul_exporter/blob/master/scripts/errcheck_excludes.txt) to pass the CI (see [here](https://github.com/prometheus/prometheus/blob/0fbc8c13a71d18c0d6c50b11d33f339954451d46/scripts/errcheck_excludes.txt#L9-L10)).

**Comment by simonpasquier** [comment]
@sowmiyamuthuraman can you rebase on master and update the `scripts/errcheck_excludes.txt` file?

**Comment by sowmiyamuthuraman** [comment]
@simonpasquier  i resolved the conflicts and updated the scripts/errcheck_excludes.txt. Can you please help in passing the test?

**Comment by simonpasquier** [comment]
I've pushed an update to your branch to fix the `go.mod` file and remove the `.idea` directory. Make sure you update your local directory.

Tests are failing because `Exporter` depends on the global `logger` instance which isn't set in the tests.
You need to add a `log.Logger` field to the `Exporter` struct and update `NewExporter()` accordingly. Then you can pass a `log.NewNopLogger()` in the tests.

**Comment by sowmiyamuthuraman** [comment]
> I've pushed an update to your branch to fix the `go.mod` file and remove the `.idea` directory. Make sure you update your local directory.
> 
> Tests are failing because `Exporter` depends on the global `logger` instance which isn't set in the tests.
> You need to add a `log.Logger` field to the `Exporter` struct and update `NewExporter()` accordingly. Then you can pass a `log.NewNopLogger()` in the tests.

Thanks, @simonpasquier. I have made the suggested changes and amended with the same commit

File: consul_exporter.go [hunk 1]
```diff
@@ -170,6 +177,7 @@ func NewExporter(opts consulOpts, kvPrefix, kvFilter string, healthSummary bool)
 		kvPrefix:      kvPrefix,
 		kvFilter:      regexp.MustCompile(kvFilter),
 		healthSummary: healthSummary,
+		logger:        log.NewLogfmtLogger(os.Stdout),
```
Comment (simonpasquier) [inline consul_exporter.go]: You should modify the `NewExporter` function to add a `log.Logger` parameter and pass it here.

File: consul_exporter.go [hunk 2]
```diff
@@ -98,10 +102,12 @@ var (
...
 )
 
-type promHTTPLogger struct{}
+type promHTTPLogger struct {
+	logger log.Logger
+}
 
 func (l promHTTPLogger) Println(v ...interface{}) {
-	log.Error(v...)
+	level.Error(l.logger).Log("msg", v)
```
Comment (simonpasquier) [inline consul_exporter.go]: rather `level.Error(l.logger).Log("msg", fmt.Sprintf(v))`

File: consul_exporter.go [hunk 3]
```diff
@@ -441,7 +452,9 @@ func main() {
 			promhttp.HandlerFor(
 				prometheus.DefaultGatherer,
 				promhttp.HandlerOpts{
-					ErrorLog: &promHTTPLogger{},
+					ErrorLog: &promHTTPLogger{
+						logger: log.NewLogfmtLogger(os.Stderr),
```
Comment (simonpasquier) [inline consul_exporter.go]: `logger` instead of `log.NewLogfmtLogger(os.Stderr)`

File: consul_exporter.go [hunk 4]
```diff
@@ -416,23 +424,26 @@ func main() {
...
 
-	log.Infoln("Starting consul_exporter", version.Info())
-	log.Infoln("Build context", version.BuildContext())
+	level.Info(logger).Log("msg", "Starting consul_exporter", "version", version.Info())
+	level.Info(logger).Log("msg", "Build context", version.BuildContext())
 
 	exporter, err := NewExporter(opts, *kvPrefix, *kvFilter, *healthSummary)
 	if err != nil {
-		log.Fatalln(err)
+		level.Error(logger).Log("msg", "Error creating an exporter", "err", err)
```
Comment (simonpasquier) [inline consul_exporter.go]: (nit) Error creating the exporter

File: consul_exporter.go [hunk 5]
```diff
@@ -416,23 +424,26 @@ func main() {
...
+	promlogConfig := &promlog.Config{}
+	flag.AddFlags(kingpin.CommandLine, promlogConfig)
 	kingpin.HelpFlag.Short('h')
 	kingpin.Parse()
+	logger := promlog.New(promlogConfig)
 
-	log.Infoln("Starting consul_exporter", version.Info())
-	log.Infoln("Build context", version.BuildContext())
+	level.Info(logger).Log("msg", "Starting consul_exporter", "version", version.Info())
+	level.Info(logger).Log("msg", "Build context", version.BuildContext())
```
Comment (simonpasquier) [inline consul_exporter.go]: Arguments to `Log()` should always be even (key+value) so:

`level.Info(logger).Log("build_context", version.BuildContext())`

File: consul_exporter.go [hunk 6]
```diff
@@ -330,14 +338,14 @@ func (e *Exporter) collectHealthSummary(ch chan<- prometheus.Metric, serviceName
 func (e *Exporter) collectOneHealthSummary(ch chan<- prometheus.Metric, serviceName string) {
 	// See https://github.com/hashicorp/consul/issues/1096.
 	if strings.HasPrefix(serviceName, "/") {
-		log.Warnf("Skipping service %q because it starts with a slash", serviceName)
+		level.Warn(e.logger).Log("msg", "Skipping service because it starts with a slash", "service name", serviceName)
```
Comment (simonpasquier) [inline consul_exporter.go]: "service name" -> "service_name"

File: consul_exporter.go [hunk 7]
```diff
@@ -107,7 +107,7 @@ type promHTTPLogger struct {
 }
 
 func (l promHTTPLogger) Println(v ...interface{}) {
-	level.Error(l.logger).Log("msg", v)
+	level.Error(l.logger).Log("msg", fmt.Sprintf("%v", v))
```
Comment (simonpasquier) [inline consul_exporter.go]: Sorry my previous comment was confusing, it should be `level.Error(l.logger).Log("msg", fmt.Sprint(v))`

File: consul_exporter.go [hunk 8]
```diff
@@ -98,10 +102,12 @@ var (
...
 )
 
-type promHTTPLogger struct{}
+type promHTTPLogger struct {
+	logger log.Logger
+}
 
 func (l promHTTPLogger) Println(v ...interface{}) {
-	log.Error(v...)
+	level.Error(l.logger).Log("msg", fmt.Sprintf("%v", v))
```
Comment (simonpasquier) [inline consul_exporter.go]: Sorry my previous comment was confusing, it should be `level.Error(l.logger).Log("msg", fmt.Sprint(v))`

**Comment by simonpasquier** [comment]
As https://github.com/prometheus/consul_exporter/pull/145 got merged, you'd need to rebase on master and adjust the changes. Sorry for the incovenience!

**Comment by simonpasquier** [comment]
Thanks for the work!

**Pull request merged**